src/org/rascalmpl/library/lang/paths/Windows.rsc

```diff
@@ -0,0 +1,206 @@
+}
+loc parseWindowsPath(str input, loc src=|unknown:///|) = mapPathToLoc(parse(#WindowsPath, input, src));
+
+@synopsis{UNC}
+private loc mapPathToLoc((WindowsPath) `<Slash _><Slash _><Slashes? _><PathChar* hostName><Slashes _><PathChar* shareName><Slashes _><WindowsFilePath path>`)
+    = appendPath(|unc://<hostName>/| + "<shareName>", path);
+
+@synopsis{DOS UNC Device Drive}
+private loc mapPathToLoc((WindowsPath) `<Slash _><Slash _><Slashes? _><DOSDevice dq><Slashes _><Drive drive>:<OptionalWindowsFilePath path>`)
+    = appendPath(|unc://<deviceIndicator(dq)>/| + "<drive>:", path);
+
+@synopsis{DOS UNC Device Path}
+private loc mapPathToLoc((WindowsPath) `<Slash _><Slash _><Slashes? _><DOSDevice dq><Slashes _><PathChar* deviceName><Slashes _><WindowsFilePath path>`)
+    = appendPath(|unc://<deviceIndicator(dq)>/| + "<deviceName>", path);
+
+private str deviceIndicator((DOSDevice) `?`) = "%3F";
+private str deviceIndicator((DOSDevice) `.`) = "%2E";
+
+@synopsis{DOS UNCPath}
+private loc mapPathToLoc((WindowsPath) `<Slash _><Slash _><Slashes? _>?<Slashes _><PathChar* shareName><Slashes _><WindowsFilePath path>`)
+    = appendPath(|unc://%3F/| + "<shareName>", path);
+
+
+@synopsis{Absolute: given the drive and relative to its root.}
+private loc mapPathToLoc((WindowsPath) `<Drive drive>:<Slashes _><WindowsFilePath path>`) 
+    = appendPath(|file:///<drive>:/|, path);
+
+@synopsis{Drive relative: relative to the current working directory on the given drive.}
+private loc mapPathToLoc((WindowsPath) `<Drive drive>:<WindowsFilePath path>`) 
+    = appendPath(|file:///<drive>:.|, path);
+
+@synopsis{Directory relative: relative to the root of the current drive.}
+private loc mapPathToLoc((WindowsPath) `<Slash _><WindowsFilePath path>`) 
+    = appendPath(|cwdrive:///|, path);
+
+@synopsis{Relative to the current working directory on the current drive.}
+private loc mapPathToLoc((WindowsPath) `<WindowsFilePath path>`) 
+    = appendPath(|cwd:///|, path);
+
+private loc appendPath(loc root, WindowsFilePath path)
+    = (root | it + "<segment>" | segment <- path.segments);
+
+private loc appendPath(loc root, (OptionalWindowsFilePath) ``) = root;
+
+private loc appendPath(loc root, (OptionalWindowsFilePath) `<Slashes _><WindowsFilePath path>`) 
+    = appendPath(root, path);
+
+private bool IS_WINDOWS = /win/i := getSystemProperty("os.name");
+
+test bool uncSharePath()
+    = parseWindowsPath("\\\\Server2\\Share\\Test\\Foo.txt")
+    == |unc://Server2/Share/Test/Foo.txt|;
+
+test bool uncDrivePath()
+    = parseWindowsPath("\\\\system07\\C$\\")
+    == |unc://system07/C$|;
+
+
+test bool uncDOSDevicePathLocalFileQuestion() {
+    loc l = parseWindowsPath("\\\\?\\c:");
+    
+    if (IS_WINDOWS) {
+        assert exists(l);
+    }
+
+    return l == |unc://%3F/c:|;
+}
+
+test bool uncDOSDevicePathLocalFileDot() {
+    loc l = parseWindowsPath("\\\\.\\C:\\Test\\Foo.txt");
+
+    
+    if (IS_WINDOWS) {
+        assert exists(l);
```

Comment:
why would we assume that `c:\test\foo.txt` is a file that exists?

Reply:
that makes no sense indeed. copy/paste error.

Reply:
the C:\ should exist however, but I have not found the reason why that fails yet. Currently trying to install Windows in VirtualBox.